ddtrace/tracer/spancontext.go

```diff
@@ -326,11 +326,30 @@ func (t *trace) finishedOne(s *span) {
 	atomic.AddInt64(&tr.spansFinished, int64(len(t.spans)))
 	sd := samplingDecision(atomic.LoadInt64((*int64)(&t.samplingDecision)))
 	if sd != decisionKeep {
+		// if trace sampling decision is drop, we still want to send single spans
+		// unless there are no single span sampling rules defined
+		var singleSpans []*span
+		if tr.rulesSampling.HasSpanRules() {
+			canDropP0s := tr.config.canDropP0s()
+			for _, span := range t.spans {
+				if tr.rulesSampling.sampleSpan(span) {
+					// since stats are computed on the tracer side, the keyTopLevel tag
+					// must be removed to preserve stats correctness
+					if canDropP0s {
+						delete(span.Metrics, keyTopLevel)
+					}
+					singleSpans = append(singleSpans, span)
+				}
+			}
+		}
 		if p, ok := t.samplingPriorityLocked(); ok && p == ext.PriorityAutoReject {
-			atomic.AddUint64(&tr.droppedP0Spans, uint64(len(t.spans)))
+			atomic.AddUint64(&tr.droppedP0Spans, uint64(len(t.spans)-len(singleSpans)))
 			atomic.AddUint64(&tr.droppedP0Traces, 1)
 		}
```

Comment:
Open question: if a trace's decision is drop, but it contains some single spans which are sent to the agent, should the tr.droppedP0Traces be increased or not?

cc @gbbr @dgoffredo

Reply:
If I'm not mistaking this was purely used for telemetry, so maybe adding just another tag `partial: true` should be enough to differentiate between fully dropped and partially dropped traces?